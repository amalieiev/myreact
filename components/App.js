export default function (props) {
  return `
        <h2>App Component</h2>
    `;
}
