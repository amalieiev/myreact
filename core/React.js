let globalParent;
let globalId = 0;
const callbacks = new Map();
const states = new Map();

export function render(component, props, parent) {
    const state = states.get(parent);
    states.set(parent, state ? state : []);

    globalParent = parent;

    const HTML = component(props);

    parent.innerHTML = HTML;

    const callback = callbacks.get(parent);

    if (callback) callback(parent);

    globalId = 0;
}

export function createSubject(initialValue) {
    return {
        value: initialValue,
        callbacks: [],
        subscribe(callback) {
            this.callbacks.push(callback);
        },
        next(value) {
            this.value = value;
            this.callbacks.forEach((callback) => callback(value));
        },
    };
}

export function useSubject(initialValue) {
    return ((id, parent) => {
        const state = states.get(parent);

        if (state[id] === undefined) {
            state[id] = createSubject(initialValue);
        }

        globalId++;

        return state[id];
    })(globalId, globalParent);
}

export function useRendered(callback) {
    return ((parent) => {
        callbacks.set(parent, callback);
    })(globalParent);
}
