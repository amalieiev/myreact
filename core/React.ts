let globalParent: HTMLElement;
let globalId: number = 0;
const callbacks = new Map<HTMLElement, () => () => void>();
const rollbacks = new Map<HTMLElement, () => void>();
const states = new Map<HTMLElement, Array<Subject<any>>>();

export type FC = (props: Record<string, any>) => string;

export class Subject<T> {
    public value: T;

    private callbacks: Array<() => void>;

    public constructor(initialValue: T) {
        this.value = initialValue;
    }

    public subscribe(callback: () => void): void {
        this.callbacks.push(callback);
    }

    public unsubscribe(callback: () => void): void {
        this.callbacks = this.callbacks.filter((item) => item !== callback);
    }

    public next(value: T): void {
        this.value = value;
        this.callbacks.forEach((callback) => {
            callback();
        });
    }
}

export function createSubject<T>(initialValue: T): Subject<T> {
    return new Subject<T>(initialValue);
}

export function render(component: FC, props: any, parent: HTMLElement): void {
    const state = states.get(parent);
    const rollback = rollbacks.get(parent);

    states.set(parent, state ? state : []);

    globalParent = parent;

    if (rollback) {
        rollback();
    }

    const HTML = component(props);

    parent.innerHTML = HTML;

    const callback = callbacks.get(parent);

    if (callback) {
        rollbacks.set(parent, callback());
    }

    globalId = 0;
}

export function useSubject<T>(initialValue: T): Subject<T> {
    return ((id: number, parent: HTMLElement): Subject<T> => {
        const state: Array<Subject<T>> = states.get(parent) as Array<
            Subject<T>
        >;

        if (state[id] === undefined) {
            state[id] = createSubject<T>(initialValue);
        }

        globalId++;

        return state[id];
    })(globalId, globalParent);
}

export function useParent(): HTMLElement {
    return ((parent: HTMLElement): HTMLElement => {
        return parent;
    })(globalParent);
}

export function useRendered(callback: () => () => void): void {
    return ((parent: HTMLElement): void => {
        callbacks.set(parent, callback);
    })(globalParent);
}
