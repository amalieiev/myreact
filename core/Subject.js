export class Subject {
    constructor(value) {
        this.value = value;
        this.subscribers = [];
    }

    subscribe(callback) {
        this.subscribers.push(callback);
    }

    current() {
        return this.value;
    }

    next(value) {
        this.value = value;
        this.subscribers.forEach((callback) => {
            callback(value);
        });
    }
}
