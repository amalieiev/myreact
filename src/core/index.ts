let globalParent: HTMLElement;
let globalId: number = 0;
const callbacks = new Map<HTMLElement, void | (() => void)>();
const rollbacks = new Map<HTMLElement, void | (() => void)>();
const states = new Map<HTMLElement, Array<Subject<any>>>();

export type FC = (props: Record<string, any>) => string;

export type HTMLElementEvent<T extends HTMLElement> = Event & {
    target: T;
};

export class Subject<T> {
    public value: T;

    private callbacks: Array<() => void> = [];

    public constructor(initialValue: T) {
        this.value = initialValue;
    }

    public subscribe(callback: () => void): void {
        this.callbacks.push(callback);
    }

    public unsubscribe(callback: () => void): void {
        this.callbacks = this.callbacks.filter((item) => item !== callback);
    }

    public next(value: T): void {
        this.value = value;
        this.callbacks.forEach((callback) => {
            callback();
        });
    }
}

export function createSubject<T>(initialValue: T): Subject<T> {
    return new Subject<T>(initialValue);
}

export function render(component: FC, props: any, parent: HTMLElement): void {
    console.log(parent);

    const state = states.get(parent);
    const rollback = rollbacks.get(parent);

    states.set(parent, state ? state : []);

    globalParent = parent;

    if (rollback) {
        rollback();
    }

    const HTML = component(props);

    console.log(HTML);

    parent.innerHTML = HTML;

    const callback = callbacks.get(parent);

    if (callback) {
        rollbacks.set(parent, callback());
    }

    globalId = 0;
}

export function useSubject<T>(initialValue: T): Subject<T> {
    return ((id: number, parent: HTMLElement): Subject<T> => {
        const state: Array<Subject<T>> = states.get(parent) as Array<Subject<T>>;

        if (state[id] === undefined) {
            state[id] = createSubject<T>(initialValue);
        }

        globalId++;

        return state[id];
    })(globalId, globalParent);
}

export function useParent(): HTMLElement {
    return ((parent: HTMLElement): HTMLElement => {
        return parent;
    })(globalParent);
}

export function useRendered(callback: () => void | (() => void)): void {
    return ((parent: HTMLElement): void => {
        callbacks.set(parent, callback);
    })(globalParent);
}
