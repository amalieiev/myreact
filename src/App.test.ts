// src/__tests__/example.js
// query utilities:
import {
    getByLabelText,
    getByText,
    getByTestId,
    queryByTestId,
    fireEvent,
    queries,
    waitFor,
    queryByText,
} from "@testing-library/dom";
// adds special assertions like toHaveTextContent
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";

import { App } from "./App";
import { render } from "./core";

test("testing App component", async () => {
    const container = document.createElement("div");

    render(App, {}, container);

    expect(getByText(container, /Type Text/)).toBeTruthy();

    const input = getByLabelText(container, /Type Text/);

    // userEvent.type(input, "Hello World");

    fireEvent.keyUp(input, { target: { value: "H" } });
    fireEvent.keyUp(input, { target: { value: "He" } });
    fireEvent.keyUp(input, { target: { value: "Hel" } });
    fireEvent.keyUp(input, { target: { value: "Hell" } });
    fireEvent.keyUp(input, { target: { value: "Hello" } });

    // setTimeout(() => {
    //     container.querySelector("p").innerHTML = "Hello World";
    // }, 100);

    expect(container).toMatchSnapshot();

    await waitFor(() => expect(queryByText(container, /Hello/)).toBeTruthy());
});
