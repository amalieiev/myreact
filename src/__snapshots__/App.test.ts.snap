// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing App component 1`] = `
<div>
  
        
  <label
    for="username"
  >
    
            
    <input
      id="username"
      type="text"
      value=""
    />
    
            Type Text
        
  </label>
  
        
  <p
    data-testid="output"
  >
    <p>
      Hello
    </p>
  </p>
  
    
</div>
`;
